File: backend/profile_search_preferences.py
"""Resolve job-search role/location from profile fields and onboarding extras."""
from __future__ import annotations

from typing import Any, Dict, List, Optional


def _first_non_empty_string(*values: Any) -> str:
    for value in values:
        if value is None:
            continue
        text = str(value).strip()
        if text:
            return text
    return ""


def onboarding_from_profile(profile: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    return dict((profile or {}).get("extras") or {}).get("onboarding") or {}


def _first_role(profile: Dict[str, Any], onboarding: Dict[str, Any]) -> str:
    for roles in (profile.get("target_roles"), onboarding.get("selected_roles")):
        if isinstance(roles, list):
            for role in roles:
                text = _first_non_empty_string(role)
                if text:
                    return text
    return ""


def resolve_profile_target_role(profile: Optional[Dict[str, Any]]) -> str:
    profile = profile or {}
    onboarding = onboarding_from_profile(profile)
    return _first_non_empty_string(profile.get("target_role"), _first_role(profile, onboarding))


def resolve_profile_target_location_data(profile: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    profile = profile or {}
    onboarding = onboarding_from_profile(profile)
    contact = profile.get("contact") or {}
    for candidate in (
        profile.get("target_location_data"),
        contact.get("location_data"),
        onboarding.get("onboarding_location_data"),
    ):
        if isinstance(candidate, dict) and candidate:
            return dict(candidate)
    return {}


def resolve_profile_target_location_label(profile: Optional[Dict[str, Any]]) -> str:
    profile = profile or {}
    onboarding = onboarding_from_profile(profile)
    contact = profile.get("contact") or {}
    location_data = resolve_profile_target_location_data(profile)
    return _first_non_empty_string(
        profile.get("target_location"),
        contact.get("location"),
        onboarding.get("onboarding_location"),
        location_data.get("location_label"),
    )

File: frontend/src/lib/profileSearchPreferences.js
import { api } from "./api";
import { EXPERIENCE_LEVELS } from "../components/onboarding/onboardingData";

/** Read search role/location from profile fields with onboarding extras fallback. */
export function resolveProfileSearchPreferences(profile) {
  const onboarding = profile?.extras?.onboarding || {};
  const contact = profile?.contact || {};
  const role = (
    profile?.target_role ||
    profile?.target_roles?.[0] ||
    onboarding.selected_roles?.[0] ||
    ""
  ).trim();
  const locationData =
    profile?.target_location_data ||
    contact?.location_data ||
    onboarding.onboarding_location_data ||
    null;
  const location = (
    profile?.target_location ||
    contact?.location ||
    onboarding.onboarding_location ||
    locationData?.location_label ||
    ""
  ).trim();
  return { role, location, locationData };
}

/** Promote onboarding city/role answers to profile search preferences. */
export async function syncOnboardingSearchPreferences({
  selectedRoles = [],
  onboardingLocation = "",
  onboardingLocationData = null,
  contractType,
  experienceId,
} = {}) {
  const roles = Array.isArray(selectedRoles)
    ? selectedRoles.map((role) => String(role || "").trim()).filter(Boolean)
    : [];
  const primaryRole = roles[0] || "";
  const locationLabel = (onboardingLocationData?.location_label || onboardingLocation || "").trim();

  if (!primaryRole && !locationLabel) return;

  const exp = EXPERIENCE_LEVELS.find((entry) => entry.id === experienceId);
  const payload = { remote_preference: "any" };
  if (primaryRole) {
    payload.target_role = primaryRole;
    payload.target_roles = roles.length ? roles : [primaryRole];
  }
  if (locationLabel) {
    payload.target_location = locationLabel;
    payload.target_location_data = onboardingLocationData || null;
  }
  if (contractType) payload.contract_type = contractType;
  if (exp?.backend) payload.seniority = exp.backend;

  await api.put("/profile/preferences", payload);

  if (locationLabel) {
    try {
      await api.put("/profile/contact", {
        location: locationLabel,
        location_data: onboardingLocationData || undefined,
      });
    } catch (_) {
      /* preferences are enough for feed fallback */
    }
  }
}

export function onboardingSnapshotToSearchSync(snapshot = {}) {
  return {
    selectedRoles: snapshot.selected_roles || [],
    onboardingLocation: snapshot.onboarding_location || "",
    onboardingLocationData: snapshot.onboarding_location_data || null,
    contractType: snapshot.contract_type,
    experienceId: snapshot.experience,
  };
}
